stdlib/LinearAlgebra/src/symmetric.jl

```diff
@@ -856,3 +875,49 @@ for func in (:log, :sqrt)
         end
     end
 end
+
+"""
+    hermitianpart(A, uplo=:U) -> Hermitian
+
+Return the Hermitian part of the square matrix `A`, defined as `(A + A') / 2`, as a
+[`Hermitian`](@ref) matrix. For real matrices `A`, this is also known as the symmetric part
+of `A`. The optional argument `uplo` controls the corresponding argument of the
+[`Hermitian`](@ref) view. For real matrices, the latter is equivalent to a
+[`Symmetric`](@ref) view.
+
+See also [`hermitianpart!`](@ref) for the corresponding in-place operation.
+
+!!! compat "Julia 1.10"
+    This function requires Julia 1.10 or later.
+"""
+hermitianpart(A::AbstractMatrix{T}, uplo::Symbol=:U) where {T} =
+    hermitianpart!(copy_similar(A, Base.promote_op(/, T, Int)), uplo)
+hermitianpart(x::Number, uplo::Symbol=:U) = real(x)
+
+"""
+    hermitianpart!(A, uplo=:U) -> Hermitian
+
+Overwrite the square matrix `A` with its Hermitian part `(A + A') / 2`, and return
+[`Hermitian(A, uplo)`](@ref). For real matrices `A`, this is also known as the symmetric
+part of `A`.
+
+See also [`hermitianpart`](@ref).
+
+!!! compat "Julia 1.10"
+    This function requires Julia 1.10 or later.
+"""
+hermitianpart!(A::AbstractMatrix, uplo::Symbol=:U) =
+    Hermitian(_hermorsympart!(a -> hermitianpart(a, uplo), adjoint, A), uplo)
+
+@inline function _hermorsympart!(real::Function, conj::Function, A::AbstractMatrix)
+    require_one_based_indexing(A)
+    n = checksquare(A)
+    @inbounds for j in 1:n
+        A[j, j] = real(A[j, j])
+        for i in 1:j-1
+            A[i, j] = val = (A[i, j] + conj(A[j, i])) / 2
+            A[j, i] = conj(val)
+        end
+    end
+    return A
+end
```

Comment:
This function is only used in one place with one set of functions as input so perhaps it's not needed and can just be the body of the `hermitianpart!(::AbstractMatrix)` method.

```suggestion
function hermitianpart!(A::AbstractMatrix, uplo::Symbol=:U)
    require_one_based_indexing(A)
    n = checksquare(A)
    @inbounds for j in 1:n
        A[j, j] = hermitianpart(A[j, j])
        for i in 1:(j - 1)
            A[i, j] = val = (A[i, j] + adjoint(A[j, i])) / 2
            A[j, i] = adjoint(val)
        end
    end
    return Hermitian(A, uplo)
end
```

But if we're returning a `Hermitian`-wrapped matrix, presumably we don't need to set both `A[i, j]` and `A[j, i]` explicitly?

Reply:
I went with @stevengj's proposal above. Since setting both entries doesn't double the runtime, we agreed on always doing it so people can take the underlying matrix as a numerically Hermitian matrix, without creating yet another copy of the `Hermitian` view.